[styled-components] Add set 'children' prop on StyledComponentProps

devrelm opened the pull request:
I had trouble when trying to use a component that had been styled whose props were all optional. I tracked it down to `children` not being added as a prop anywhere.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).
- [ ] N/A ~Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>~
- [ ] N/A ~Increase the version number in the header if appropriate.~
- [ ] N/A ~If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.~

Kovensky commented on types/styled-components/test/index.tsx:
```diff
@@ -25,6 +25,67 @@ import {} from "styled-components/cssprop";
+// a "no properties in common" error.
+const UseStyledAllOptionalPropComponent: React.SFC = () => (
+    <StyledAllOptionalPropComponent>
+        blah
```
There's a bug in TS that doesn't actually trigger excess property errors with `children` in certain cases, so this would also need an `$ExpectError` equivalent.

Don't define such a component with `React.SFC`, use a function declaration as it's the only thing not directly poisoned by the automatic `{ children?: React.ReactNode }` which must be killed.

Kovensky replied:
Actually, since you did declare `AllOptionalPropComponent` with `React.SFC`, it _should_ have an implicit `{ children?: React.ReactNode }` in it. I wonder if the `&` in the `React.SFC` is erasing it from the extracted types in `React.ComponentProps`.

Yet one more reason why the automatic `{ children?: React.ReactNode }` must be killed.

Kovensky commented:
This PR that I never got around to finishing is related and might have fixed the root cause of this particular bug: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/30913

devrelm commented:
@Igorbek The problem actually originates not as the component failing to accept children, but as the component failing to accept _only_ children when the Props type has only optional properties declared. This use-case is covered by the first test case I added, using the `AllOptionalProps` type. I would have thought that passing _only_ `children` to a styled component would happen more often than passing _no_ `children` to a styled component, but the only data I have to back that up is my own codebase, and I admit that it could be an outlier.

Still, I do agree in general with @Kovensky that (if I'm not putting words in their mouth) `children` would make more sense as an explicitly-declared prop rather than being implicitly added by the `react` types, but I imagine that that would be a pretty epic breaking change at this point.

I had already gone ahead and added the `children` prop directly to the props type in my own code to correct the error. I originally intended it as a temporary workaround until this PR was approved, but I suppose it might have to hang in there a little longer.

@Kovensky Are there any (small/non-breaking) changes you could see being made that could fix this, even if it means updating the react types instead of styled-components? If not, I can go ahead and close this PR.

Also, who do we petition about no longer auto-adding `children` to props types? It's not like it's gonna get any easier as React and TypeScript continue to gain followers, so if we're gonna do it we should have a discussion on how to best do it and get the ball rolling.

Kovensky commented:
The more immediate fix is to change `type ComponentProps` to use `extends ((props: infer P) => ReactNode) | (new (props: infer P) => Component<any>))` instead of `extends ComponentType<infer P>`.

Basically, replace direct or indirect references to `FunctionComponent<infer P>` with `(props: infer P) => ReactNode)`, and `ComponentClass<infer P>` with `new (props: infer P) => Component<any>`. This causes the inference to also include the `& { children?: ReactNode }` that `FunctionComponent` is adding for you, instead of removing it as redundant.

---

To be honest, I want to overhaul the `react` types but I don't want to cause a python3 situation because it'd have to be almost entirely backwards incompatible to be able to throw out all the tech debt.

If I figure this out I need to do some research on codemods first, and I'll probably need help from DT as an org to be able to push this through.

Perhaps I just should write down what I have in mind (similar to https://github.com/Microsoft/TypeScript/issues/28954) and then we can explore what to do from there.

devrelm commented:
Sounds good. I might not get to it until tonight or tomorrow (or even later in the week, to be honest), but I'll implement your suggestions and update this PR if it works out or report back here if I hit any roadblocks.
